Some worrying warnings in 0.86

We have a number of warnings building 0.86. 

Some are harmless (but we should deal with them - why implicit declarations?), some definitely look not harmless at all (incompatible pointer points) and some happen in libGPAC which we just updated and that we are not going to fix directly in CCExtractor (but if someone wants to contribute to libGPAC I'm sure they'd appreciate it).

Warnings we should fix follow:

../src/lib_ccx/ccx_encoders_transcript.c: In function ‘write_cc_bitmap_as_transcript’:
../src/lib_ccx/ccx_encoders_transcript.c:45:6: warning: implicit declaration of function ‘millis_to_date’ [-Wimplicit-function-declaration]
      millis_to_date(start_time + context->subs_delay, buf1, context->date_format, context->millis_separator);
      ^~~~~~~~~~~~~~
../src/lib_ccx/ccx_encoders_transcript.c: In function ‘write_cc_subtitle_as_transcript’:
../src/lib_ccx/ccx_encoders_transcript.c:215:5: warning: implicit declaration of function ‘mprint’ [-Wimplicit-function-declaration]
     mprint("Warning:Loss of data\n");
     ^~~~~~
../src/lib_ccx/general_loop.c: In function ‘rcwt_loop’:
../src/lib_ccx/general_loop.c:1252:28: warning: passing argument 1 of ‘telxcc_dump_prev_page’ from incompatible pointer type
-Wincompatible-pointer-types]
      telxcc_dump_prev_page(telctx,dec_sub);
                            ^~~~~~
In file included from ../src/lib_ccx/lib_ccx.h:22:0,
                 from ../src/lib_ccx/general_loop.c:1:
../src/lib_ccx/teletext.h:106:6: note: expected ‘struct TeletextCtx *’ but argument is of type ‘struct Teletext *’
 void telxcc_dump_prev_page(struct TeletextCtx *ctx, struct cc_subtitle *sub);
      ^~~~~~~~~~~~~~~~~~~~~
../src/lib_ccx/ccx_encoders_common.c: In function ‘write_subtitle_file_footer’:
../src/lib_ccx/ccx_encoders_common.c:374:4: warning: implicit declaration of function ‘write_spumux_footer’ [-Wimplicit-function-declaration]
    write_spumux_footer(out);
    ^~~~~~~~~~~~~~~~~~~
../src/lib_ccx/ccx_encoders_common.c: In function ‘write_subtitle_file_header’:
../src/lib_ccx/ccx_encoders_common.c:528:4: warning: implicit declaration of function ‘write_spumux_header’ [-Wimplicit-function-declaration]
    write_spumux_header(ctx, out);
    ^~~~~~~~~~~~~~~~~~~
../src/lib_ccx/ocr.c: In function ‘ocr_rect’:
../src/lib_ccx/ocr.c:762:3: warning: implicit declaration of function ‘mapclut_paletee’ [-Wimplicit-function-declaration]
   mapclut_paletee(palette, alpha, (uint32_t *)rect->data[1],rect->nb_colors);
   ^~~~~~~~~~~~~~~
../src/lib_ccx/ccx_encoders_spupng.c: In function ‘utf8_to_utf32’:
../src/lib_ccx/ccx_encoders_spupng.c:813:9: warning: return from incompatible pointer type [-Wincompatible-pointer-types]
  return string_utf32;
         ^~~~~~~~~~~~
In file included from ../src/lib_ccx/ffmpeg_intgr.c:1:0:
../src/lib_ccx/ffmpeg_intgr.h:17:33: warning: ‘struct lib_ccx_ctx’ declared inside parameter list will not be visible outside
of this definition or declaration
 int ffmpeg_get_more_data(struct lib_ccx_ctx *ctx, struct demuxer_data **ppdata);
                                 ^~~~~~~~~~~

--- a/src/lib_ccx/ccx_encoders_common.h
+++ b/src/lib_ccx/ccx_encoders_common.h
@@ -207,6 +207,9 @@ int write_cc_bitmap_as_spupng          (struct cc_subtitle *sub, struct encoder_
 int write_cc_bitmap_as_transcript      (struct cc_subtitle *sub, struct encoder_ctx *context);
 int write_cc_bitmap_as_libcurl         (struct cc_subtitle *sub, struct encoder_ctx *context);
 
+void write_spumux_header(struct encoder_ctx *ctx, struct ccx_s_write *out);
+void write_spumux_footer(struct ccx_s_write *out);
+
 struct cc_subtitle * reformat_cc_bitmap_through_sentence_buffer (struct cc_subtitle *sub, struct encoder_ctx *context);
 
 void set_encoder_last_displayed_subs_ms(struct encoder_ctx *ctx, LLONG last_displayed_subs_ms);

--- a/src/lib_ccx/ccx_encoders_helpers.h
+++ b/src/lib_ccx/ccx_encoders_helpers.h
@@ -6,6 +6,7 @@
 #include "ccx_decoders_structs.h"
 #include "ccx_decoders_608.h"
 #include "ccx_encoders_common.h"
+#include "png.h"
 
 extern char **spell_lower;
 extern char **spell_correct;
@@ -40,4 +41,6 @@ void shell_sort(void *base, int nb, size_t size, int(*compar)(const void*p1, con
 
 void ccx_encoders_helpers_perform_shellsort_words(void);
 void ccx_encoders_helpers_setup(enum ccx_encoding_type encoding, int no_font_color, int no_type_setting, int trim_subs);
+
+int mapclut_paletee(png_color *palette, png_byte *alpha, uint32_t *clut, uint8_t depth);
 #endif

--- a/src/lib_ccx/ccx_encoders_spupng.c
+++ b/src/lib_ccx/ccx_encoders_spupng.c
@@ -799,11 +799,11 @@ uint32_t* utf8_to_utf32(char* src) {
 	len_src = strlen(src);
 	len_dst = (len_src + 2) * 4; // one for FEFF and one for \0
 
-	char* string_utf32 = (char *)calloc(len_dst, 1);
+	uint32_t* string_utf32 = (char *)calloc(len_dst, 1);
 	size_t inbufbytesleft = len_src;
 	size_t outbufbytesleft = len_dst;
 	char* inbuf = src;
-	char* outbuf = string_utf32;
+	uint32_t* outbuf = string_utf32;
 
 	iconv_t cd = iconv_open("UTF-32", "UTF-8");
 	int result = iconv(cd, &inbuf, &inbufbytesleft, &outbuf, &outbufbytesleft);

--- a/src/lib_ccx/ccx_encoders_spupng.c
+++ b/src/lib_ccx/ccx_encoders_spupng.c
@@ -799,11 +799,11 @@ uint32_t* utf8_to_utf32(char* src) {
 	len_src = strlen(src);
 	len_dst = (len_src + 2) * 4; // one for FEFF and one for \0
 
-	uint32_t* string_utf32 = (char *)calloc(len_dst, 1);
+	uint32_t* string_utf32 = (uint32_t *) calloc(len_dst, 1);
 	size_t inbufbytesleft = len_src;
 	size_t outbufbytesleft = len_dst;
 	char* inbuf = src;
-	uint32_t* outbuf = string_utf32;
+	char* outbuf = (char*) string_utf32;
 
 	iconv_t cd = iconv_open("UTF-32", "UTF-8");
 	int result = iconv(cd, &inbuf, &inbufbytesleft, &outbuf, &outbufbytesleft);

--- a/src/lib_ccx/ccx_encoders_transcript.c
+++ b/src/lib_ccx/ccx_encoders_transcript.c
@@ -7,6 +7,7 @@
 #include "ccx_decoders_708_output.h"
 #include "ccx_encoders_xds.h"
 #include "ccx_encoders_helpers.h"
+#include "lib_ccx.h"
 
 #ifdef ENABLE_SHARING
 #include "ccx_share.h"

--- a/src/lib_ccx/ffmpeg_intgr.c
+++ b/src/lib_ccx/ffmpeg_intgr.c
@@ -13,9 +13,6 @@
 #include <libavutil/log.h>
 #include <libavutil/error.h>
 
-#include "lib_ccx.h"
-#include "ccx_demuxer.h"
-
 struct ffmpeg_ctx
 {
 	AVFormatContext *ifmt;

--- a/src/lib_ccx/ffmpeg_intgr.c
+++ b/src/lib_ccx/ffmpeg_intgr.c
@@ -14,6 +14,8 @@
 #include <libavutil/error.h>
 
 #include "lib_ccx.h"
+#include "ccx_demuxer.h"
+
 struct ffmpeg_ctx
 {
 	AVFormatContext *ifmt;

--- a/src/lib_ccx/ffmpeg_intgr.h
+++ b/src/lib_ccx/ffmpeg_intgr.h
@@ -1,6 +1,7 @@
 #ifndef _FFMPEG_INTIGRATION
 #define _FFMPEG_INTIGRATION
 
+#include "lib_ccx.h"
 #ifdef ENABLE_FFMPEG
 #include "libavutil/common.h"
 #include "libavutil/error.h"

--- a/src/lib_ccx/general_loop.c
+++ b/src/lib_ccx/general_loop.c
@@ -1194,7 +1194,7 @@ int rcwt_loop(struct lib_ccx_ctx *ctx)
 	int caps = 0;
 	LLONG result;
 	struct encoder_ctx *enc_ctx = update_encoder_list(ctx);
-	struct Teletext *telctx;
+	struct TeletextCtx *telctx;
 	// As BUFSIZE is a macro this is just a reminder
 	if (BUFSIZE < (3*0xFFFF + 10))
 		fatal (CCX_COMMON_EXIT_BUG_BUG, "In rcwt_loop: BUFSIZE too small for RCWT caption block.\n");

